df_py/volume/reward_calculator.py

```diff
@@ -0,0 +1,479 @@
+            for nft_addr in self.nftvols_USD[chainID]:
+                nft_addrs.add(nft_addr)
+
+        return sorted(nft_addrs)
+
+    @freeze_attributes
+    @enforce_types
+    def _get_lp_addrs(self) -> List[str]:
+        """
+        @return
+          LP_addrs -- list of unique LP addrs. Order is consistent.
+        """
+        LP_addrs = set()
+        for chainID in self.stakes:
+            for nft_addr in self.stakes[chainID]:
+                for LP_addr in self.stakes[chainID][nft_addr]:
+                    LP_addrs.add(LP_addr)
+
+        return sorted(LP_addrs)
+
+    @freeze_attributes
+    @enforce_types
+    def _get_predictoors(self) -> Dict[int, List[str]]:
+        """
+        @return
+          predictoors -- dict of (chainID, list of nft_addrs)
+        """
+        chainIDs = list(self.stakes.keys())
+        predictoors: Dict[int, List[str]] = {chain_id: [] for chain_id in chainIDs}
+
+        for chain_id in DEPLOYER_ADDRS.keys():
+            predictoors[chain_id] = query_predictoor_contracts(chain_id).keys()
+
+        return predictoors
+
+
+class RewardShaper:
+    @staticmethod
+    @enforce_types
+    def flatten(rewards: Dict[int, Dict[str, float]]) -> Dict[str, float]:
+        """
+        @arguments
+          rewards -- dict of [chainID][LP_addr] : reward_float
+
+        @return
+          flats -- dict of [LP_addr] : reward_float
+        """
+        flats: Dict[str, float] = {}
+        for chainID in rewards:
+            for LP_addr in rewards[chainID]:
+                flats[LP_addr] = flats.get(LP_addr, 0.0) + rewards[chainID][LP_addr]
+
+        return flats
+
+    @staticmethod
+    def merge(*reward_dicts):
+        merged_dict = {}
+
+        for reward_dict in reward_dicts:
+            for key, value in reward_dict.items():
+                merged_dict[key] = merged_dict.get(key, 0) + value
+
+        return merged_dict
+
+
+@enforce_types
+def get_df_week_number(dt: datetime) -> int:
+    """Return the DF week number. This is used by boundRewardsByDcv().
+    There was a gap from DF4 to DF5. Since we only care about future dates,
+    don't bother to properly support this gap, just focus on future.
+    """
+    DF5_start = datetime(2022, 9, 29)  # Thu Sep 29
+    if dt < DF5_start:
+        return -1
+
+    days_offset = (dt - DF5_start).days
+    weeks_offset = days_offset // 7
+    DF_week = weeks_offset + 1 + 4
+    return DF_week
+
+
+def calc_dcv_multiplier(DF_week: int, is_predictoor: bool) -> float:
+    """
+    Calculate DCV multiplier, for use in bounding rewards_avail by DCV
+
+    @arguments
+      DF_week -- e.g. 9 for DF9
+
+    @return
+      DCV_multiplier --
+    """
+    if is_predictoor:
+        return PREDICTOOR_MULTIPLIER
+
+    return _calc_dcv_multiplier(DF_week)
+
+
+def _calc_dcv_multiplier(DF_week: int) -> float:
+    if DF_week < 9:
+        return np.inf
+
+    if 9 <= DF_week <= 28:
+        return -0.0485 * (DF_week - 9) + 1.0
+
+    return 0.001
```

Comment:
It's confusing to have two routines, with names "calc_dcv_multiplier" and "_calc_dcv_multiplier".

I recommend to merge into a single routine.

Reply:
this was done for easier testing, but I managed to find a workaround and I will merge them.